datafusion/physical-plan/src/joins/hash_join.rs

```diff
@@ -747,6 +753,97 @@ where
     Ok(())
 }
 
+// State for storing left/right side indices used for partial batch output
+// & producing ranges for adjusting indices
+#[derive(Debug, Default)]
+pub(crate) struct HashJoinStreamState {
+    // total rows in current probe batch
+    probe_rows: usize,
+    // saved probe-build indices to resume matching from
+    last_matched_indices: Option<(usize, usize)>,
+    // current iteration has been updated
+    matched_indices_updated: bool,
+    // tracking last joined probe side index seen for further indices adjustment
+    last_joined_probe_index: Option<usize>,
+    // tracking last joined probe side index seen for further indices adjustment
+    prev_joined_probe_index: Option<usize>,
+}
+
+impl HashJoinStreamState {
```

Comment:
Another way to represent a state machine like this in Rust is via an enum. The upside is that it makes it more explicit what fields are used in what states and what transitions are allowed

Something like this perhaps
```rust
enum HashJoinOutputState {
  /// The hash table is still being built
  Hashing,
  /// The hash table has been built, and a batch of `probe_rows` has is being output
  /// but nothing has been output yet
  Begin {
    probe_rows: usize
  },
  /// Have output up to `last_matched_indices`
  Output {
    // saved probe-build indices to resume matching from
    last_matched_indices: Option<(usize, usize)>,
  } 
 /// ....
}
```

Reply:
I strongly advise to make this enum, this is a common approach in other executors as well.